### PR TITLE
Biased Histogram Support for Quantile Estimation

### PR DESCRIPTION
**Pull request by chrisdennis**
This PR contains an implementation of: http://web.cs.ucla.edu/~zaniolo/papers/a13-mousavi.pdf

In addition to that paper you may also want to look at:
 * http://web.cs.ucla.edu/~zaniolo/papers/Histogram-EDBT2011-CamReady.pdf
 * http://sacan.biomed.drexel.edu/cmteditor/program/download?confname=ssdbm2013&pid=25&file=presentation
 * http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.24.7941

There is one small correction to the 'Exponential Histogram' algorithm described in that last paper where we add a tail-correction to the 1-size boxes.  Since there are no fractional boxes (½, ¼, ⅛ ...) we have to compensate by increasing the count of the one boxes.  That correction is *l* (in the paper) or `mergeThreshold - 1` (in the code) times the sum of the infinite geometric series ½, ¼, ⅛ ... (which is 1).  I believe this is what Mousavi & Zaniolo were attempting to compensate for when they had an extra one size box in their description of the algorithm.

I will note that I'm still missing javadoc from the various public constructors, and as of yet there is no coupling between this and the statistics code.  The former I want to accomplish before merging, the later is trivial and I'd rather it be done when these get consumed by M&M.

### CODE REVIEW

**Comment by mathieucarbou** [comment]
Thanks @chrisdennis :-) 
I'll have time tomorrow afternoon to have a look on my side. I will think about its usage in Ehcache for %tiles.

File: src/main/java/org/terracotta/statistics/derived/histogram/ExponentialHistogram.java [hunk 1]
```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright 2015 Terracotta, Inc..
```
Comment (henri-tremblay) [inline src/main/java/org/terracotta/statistics/derived/histogram/ExponentialHistogram.java]: 2015. And two dots

File: src/main/java/org/terracotta/statistics/derived/histogram/ExponentialHistogram.java [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+    long[] originalBoxes = boxes;
+    int[] originalInsert = insert;
+
+    int logLast = (Long.SIZE - 1) - numberOfLeadingZeros(last);
+    ExponentialHistogram other = new ExponentialHistogram(epsilon, mergeThreshold, window, logLast - 1);
+    initializeArrays(logLast - 1);
+    this.total = 0;
+    this.last = 0;
+    
+    { //extracted zero iteration
```
Comment (henri-tremblay) [inline src/main/java/org/terracotta/statistics/derived/histogram/ExponentialHistogram.java]: Can't you extract this into a sub-method?

File: src/main/java/org/terracotta/statistics/derived/histogram/Histogram.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2015 Terracotta, Inc..
```
Comment (henri-tremblay) [inline src/main/java/org/terracotta/statistics/derived/histogram/Histogram.java]: 2015. And two dots

File: src/main/java/org/terracotta/statistics/derived/histogram/ImmutableDoubleBucket.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2015 Terracotta, Inc..
```
Comment (henri-tremblay) [inline src/main/java/org/terracotta/statistics/derived/histogram/ImmutableDoubleBucket.java]: 2015. And two dots

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    this.widthError = widthError;
+  }
+
+  @Test
+  @Ignore
+  public void evaluateFlatHistogramErrors() {
+    StandardDeviation parameter = new StandardDeviation();
+    for (int i = 0; i < 1000; i++) {
+      parameter.increment(flatHistogramFit(System.nanoTime())[1]);
+      System.out.println("Flat Histogram Slope: Iteration: " + i + " S.D:" + parameter.getResult());
```
Comment (henri-tremblay) [inline src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java]: sout?

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    this.bias = (float) Math.pow(biasRange, 1.0 / bars);
+    this.bars = bars;
+
+    this.slopeError = slopeError;
+    this.centroidError = centroidError;
+    this.widthError = widthError;
+  }
+
+  @Test
+  @Ignore
```
Comment (henri-tremblay) [inline src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java]: Why ignored? (please comment)

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    double range = 1000;
+    for (int i = 0; i < 100000; i++) {
+      bsbh.event(rndm.nextDouble() * range, i);
+    }
+
+    return fit(bsbh, PolynomialCurveFitter.create(1));
+  }
+
+  @Test
+  @Ignore
```
Comment (henri-tremblay) [inline src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java]: Same here. Why ignore

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPerformanceTest.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+package org.terracotta.statistics.derived.histogram;
+
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Random;
+
+@Ignore
```
Comment (henri-tremblay) [inline src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPerformanceTest.java]: JMH?

File: src/test/java/org/terracotta/statistics/derived/histogram/ExponentialHistogramTest.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terracotta.statistics.derived.histogram;
+
+import static org.hamcrest.core.Is.is;
```
Comment (henri-tremblay) [inline src/test/java/org/terracotta/statistics/derived/histogram/ExponentialHistogramTest.java]: It's weird that the static imports are above the others

File: src/test/java/org/terracotta/statistics/derived/histogram/ExponentialHistogramTest.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2015 Terracotta, Inc..
```
Comment (henri-tremblay) [inline src/test/java/org/terracotta/statistics/derived/histogram/ExponentialHistogramTest.java]: 2015. And two dots

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPreciseTest.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
+/*
```
Comment (henri-tremblay) [inline src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPreciseTest.java]: Shouldn't we test on threaded value insertion somewhere?

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java (hunk 6 above)
Comment (chrisdennis) [reply src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java]: It's used to establish what the S.D. is on the fit - those values get fed back in to the parameters for the test.

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java (hunk 5 above)
Comment (chrisdennis) [reply src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java]: See comment on ignoring.

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java (hunk 7 above)
Comment (chrisdennis) [reply src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java]: Same thing but for the gaussian test.

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPerformanceTest.java (hunk 8 above)
Comment (chrisdennis) [reply src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPerformanceTest.java]: Probably should be... but I'm not massively bothered to change it right now.

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPreciseTest.java (hunk 11 above)
Comment (chrisdennis) [reply src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPreciseTest.java]: Threaded?  The only thing is going to need to be locked outside of this code.

File: src/test/java/org/terracotta/statistics/derived/histogram/ExponentialHistogramTest.java (hunk 9 above)
Comment (chrisdennis) [reply src/test/java/org/terracotta/statistics/derived/histogram/ExponentialHistogramTest.java]: Will fix.

File: src/main/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogram.java [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+import static java.lang.Math.nextUp;
+import static java.util.Comparator.comparingDouble;
+import static java.util.stream.Stream.of;
+
+public class BarSplittingBiasedHistogram implements Histogram<Double> {
+  private static final float DEFAULT_MAX_COEFFICIENT = 1.7f;
```
[the remark this answers is not in the document]
Comment (chrisdennis) [reply src/main/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogram.java]: Probably not much... I guess I figured I didn't need the extra accuracy here?  Happy to standardize on doubles everywhere if people want.

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java (hunk 6 above)
Comment (henri-tremblay) [reply src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramFittingTest.java]: Got it. Can you please comment in the code about how it should be used?

File: src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPreciseTest.java (hunk 11 above)
Comment (henri-tremblay) [reply src/test/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogramPreciseTest.java]: Oh! Please specify it somewhere in the javadoc. We really should start using @NotTheadSafe

**Comment by chrisdennis** [comment]
@AbfrmBlr @mathieucarbou @cljohnso any of you guys planning to look at this in the next couple of days or should I hunt for some more reviewers?

**Comment by mathieucarbou** [comment]
@chrisdennis : I'd like to have a look yes but I the next couple of days. But I think I will much more have usage questions probably so perhaps also asking them or eventually raising a PR after is fine :-) I guess other people would need that too pretty fast ?

**Pull request merged**